src/rounded_rect.rs

```diff
@@ -0,0 +1,302 @@
+
+        // project point out of the inner rectangle (positive quadrant)
+        let px = (pt.x.abs() - inside_half_width).max(0.0);
+        let py = (pt.y.abs() - inside_half_height).max(0.0);
+
+        let inside = px * px + py * py <= radius * radius;
+        if inside {
+            1
+        } else {
+            0
+        }
+    }
+
+    #[inline]
+    fn bounding_box(&self) -> Rect {
+        self.rect.bounding_box()
+    }
+
+    #[inline]
+    fn as_rounded_rect(&self) -> Option<RoundedRect> {
+        Some(*self)
+    }
+}
+
+struct RectPathIter {
+    rect: Rect,
+    radius: f64,
+    ix: usize,
+}
+
+// This is clockwise in a y-down coordinate system for positive area.
+impl Iterator for RectPathIter {
+    type Item = PathEl;
+
+    fn next(&mut self) -> Option<PathEl> {
+        self.ix += 1;
+        match self.ix {
+            1 => Some(PathEl::MoveTo(Point::new(
+                self.rect.x0,
+                self.rect.y0 + self.radius,
+            ))),
+            2 => Some(PathEl::LineTo(Point::new(
+                self.rect.x1 - self.radius,
+                self.rect.y0,
+            ))),
+            3 => Some(PathEl::LineTo(Point::new(
+                self.rect.x1,
+                self.rect.y1 - self.radius,
+            ))),
+            4 => Some(PathEl::LineTo(Point::new(
+                self.rect.x0 + self.radius,
+                self.rect.y1,
+            ))),
+            5 => Some(PathEl::ClosePath),
+            _ => None,
+        }
+    }
+}
+
+// This is clockwise in a y-down coordinate system for positive area.
+impl Iterator for RoundedRectPathIter {
+    type Item = PathEl;
+
+    fn next(&mut self) -> Option<PathEl> {
+        if self.idx > 4 {
+            return None;
+        }
+
+        // Iterate between rectangle and arc iterators.
+        // Rect iterator will start and end the path.
+        if self.idx == 0 {
+            self.idx += 1;
+            self.rect.next()
+        } else {
+            match self.arcs[self.idx - 1].next() {
+                Some(elem) => Some(elem),
+                None => {
+                    self.idx += 1;
+                    self.rect.next()
+                }
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{Circle, Point, Rect, RoundedRect, Shape};
+
+    #[test]
+    fn area() {
+        let epsilon = 1e-9;
+
+        // Extremum: 0.0 radius corner -> rectangle
+        let rect = Rect::new(0.0, 0.0, 100.0, 100.0);
+        let rounded_rect = RoundedRect::new(0.0, 0.0, 100.0, 100.0, 0.0);
+        assert!((rect.area() - rounded_rect.area()).abs() < epsilon);
+
+        // Extremum: half-size radius corner -> circle
+        let circle = Circle::new((0.0, 0.0), 50.0);
+        let rounded_rect = RoundedRect::new(0.0, 0.0, 100.0, 100.0, 50.0);
+        assert!((circle.area() - rounded_rect.area()).abs() < epsilon);
+    }
+
+    #[test]
+    fn winding() {
```

Comment:
I think if this was tested on a rect not centered at the origin, it would reveal more problems.